File: proyectos/cppcalc/parser.cpp
#include "parser.h"
#include "calcex.h"
#include <string>
#include <sstream>

Parser::Parser(istream* in) {
   scan = new Scanner(in);
}

Parser::~Parser() {
   try {
      delete scan;
   } catch (...) {}
}

AST* Parser::parse() {
   return prog();
}

AST* Parser::prog() {
   AST* result = expr();
   Token* t = scan->getToken();

   if (t->getType() != eof) {
      cout << "Syntax Error: Expected EOF, found token at column " << t->getCol() << endl;
      throw ParseError;
   }

   return result;
}

AST* Parser::expr() {
   return restExpr(term());
}

AST* Parser::restExpr(AST* e) {
   Token* t = scan->getToken();

   if (t->getType() == add) {
      return restExpr(new AddNode(e,term()));
   }

   if (t->getType() == sub)
      return restExpr(new SubNode(e,term()));

   scan->putBackToken();

   return e;
}

AST* Parser::term() {
   return restTerm(storable());
}

AST* Parser::restTerm(AST* e) {
  Token* t = scan->getToken();

  if (t->getType() == times) {
    return restTerm(new TimesNode(e, storable()));
  }

  if (t->getType() == divide) {
    return restTerm(new DivideNode(e, storable()));
  }

  scan->putBackToken();

  return e;
}

AST* Parser::storable() {
  AST* result = factor();

  Token *t = scan->getToken();

  if (t->getType() == keyword) {
    if (t->getLex() == "S") {
      return new StoreNode(result);
    }
    else {
      cout << "Parser error: Expected s found "
	   << t->getLex() << endl;
      throw ParseError;
    }
  }

  scan->putBackToken();

  return result;
}

AST* Parser::factor() {

  Token* t = scan->getToken();
  
   if (t->getType() == number) {
     istringstream in(t->getLex());
     int val;
     in >> val;
     return new NumNode(val);
   }

   if (t->getType() == keyword) {
     if (t->getLex() == "R") {
       return new RecallNode();
     }
     else {
       cout << "Parse error: Expected R found "
	    << t->getLex() << endl;
       throw ParseError;
     }
   }

   if (t->getType() == lparen) {
     AST* result = expr();
     t = scan->getToken();
     if (t->getType() != rparen) {
       cout << "Parse error: Expected ) found "
	    << endl;
       throw ParseError;
     }
     return result;
   }

   cout << "Expected Number, 'R', '('" << endl;
   throw ParseError;}
   

File: proyectos/rubycalc/lib/parser.rb
require 'ast'
require 'scanner'
require 'token'
require 'calcex'

class Parser
  def initialize istream
    @scan = Scanner.new istream
  end

  def parse
    return prog
  end

  def prog
    result = expr
    t = @scan.getToken

    unless t.type == :eof then
      print "Expected EOF. Found ", t.type, ".\n"
      raise ParseError.new
    end

    return result
  end

  def expr
    restExpr term
  end

  def restExpr e
    t = @scan.getToken

    case t.type
    when :add
      return restExpr(AddNode.new(e,term))
    when :sub
      return restExpr(SubNode.new(e,term))
    else
      @scan.putBackToken
    end
    
    return e
  end
  
  def term
    restTerm storable
  end

  def restTerm e
    t = @scan.getToken

    case t.type
    when :times
      return restTerm(TimesNode.new(e,storable))
    when :divide
      return restTerm(DivideNode.new(e,storable))
    else
      @scan.putBackToken
    end
    
    return e
  end

  def storable
    result = factor
    t = @scan.getToken
    case t.type
    when :keyword
      unless t.lex == "S"
        print "Expected S found", t.lex
        raise ParseError.new
      end
    else
      @scan.putBackToken
    end
    
    return result
  end

  def factor
    t = @scan.getToken
    case t.type
    when :number
      return NumNode.new(t.lex.to_i)
    when :keyword
      unless t.lex == "R"
        print "Expected R found", t.lex
        raise ParseError.new
      end
      return RecallNode.new
    when :lparen
      result = expr
      t = @scan.getToken
      unless t.type == :rparen
        print "Expected )"
        raise ParseError.new
      end
      return result
    else
      print "Epected number, r, ("
      raise ParseError.new
    end
  end
  
  private :prog, :expr, :restExpr, :term, :restTerm, :storable, :factor
end
